### ISSUE TITLE
sudden crash from OOM, why (stilll) no better memory management?

### ISSUE THREAD

**Issue by rdelangh** [orientechnologies/orientdb#8126]
## Expected behavior  
When the server process is allocating memory, it should have the necessary memory-management logic to avoid that it runs out of bounds and dies uncontrollably.

## Actual behavior  
Launced a command to create a new property, on a class which contains already some 100M records ('documents'). The schema for this class is set to 'strict', because we need it.
The server died.
Its logs contain the following:
```
Java HotSpot(TM) 64-Bit Server VM warning: INFO: os::commit_memory(0x00007efc13f50000, 65536, 1) failed; error='Cannot allocate memory' (errno=12)
#
# There is insufficient memory for the Java Runtime Environment to continue.
# Native memory allocation (mmap) failed to map 65536 bytes for committing reserved memory.
# An error report file with more information is saved as:
# /opt/orientdb/orientdb-community-2.2.31-SNAPSHOT/bin/hs_err_pid27812.log
[thread 139617979836160 also had an error][thread 139617981941504 also had an error][thread 139617986262784 also had an error][thread 139617980888832 also had an
error][thread 139609873745664 also had an error]
#
# Compiler replay data is saved as:
# /opt/orientdb/orientdb-community-2.2.31-SNAPSHOT/bin/replay_pid27812.log
```
This "replay_pid" logfile is attached.
[replay_pid27812.log.gz](https://github.com/orientechnologies/orientdb/files/1764522/replay_pid27812.log.gz)

**Comment by laa** [comment]
@rdelangh do you run your system on VM machine?

**Comment by rdelangh** [comment]
hi @laa , thanks for replying.
We run in a Ubuntu LXC container, so that's not what most people call a VM such as what a hypervisor would provide like VMware or VirtualBox. The virtual server is not running within a Java RTE, but is part of the global operating system and has therefor (typically) all hardware resources available from the machine.
I found that the underlying ZFS on the physical server was not yet restraint to use limited cache memory, and so not enough RAM remained for the OrientDB server. I reconfigured ZFS to use limited cache amounts, and now ODB can do its index re-builds.
Anyhow, while doing its index re-builds, I see that the ODB server is growing dramatically in terms of RAM usage (up till 57GB)
-> can this be optimised, so that it is more memory-efficient while doing indexes?


```
3) clearly it is the Java server process eating too much RAM until it dies (look at the "RES" column in the output of "top" below):
```
$ top
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
40368 orientdb  20   0 36.594g 8.786g 102056 S 290.3  7.0   8:44.31 java
...
40368 orientdb  20   0 45.144g 0.017t 694148 S 101.7 13.7  23:29.41 java
...
40368 orientdb  20   0 51.192g 0.022t 257388 S 101.0 17.8  32:26.01 java
...

...
40368 orientdb  20   0 62.962g 0.033t 131152 S 235.2 26.5  50:28.65 java
...
40368 orientdb  20   0 75.641g 0.045t 563356 S 264.8 36.3  68:14.36 java
...
40368 orientdb  20   0 82.377g 0.051t 605008 S 320.7 41.2  77:05.42 java


=> I think the software is designed to build some stack or map of all indexed data in memory, that it would clear or flush only when the (re)indexing is completed. Such a this strategy is unrealistic when you want to position the software in the category of BIG databases.
